core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/datetime/DateTimeFunction.java

```diff
@@ -123,19 +392,167 @@ private ExprValue exprDate(ExprValue exprValue) {
+   * @return ExprValue.
+   */
+  private ExprValue exprDayOfWeek(ExprValue date) {
+    if (date instanceof ExprStringValue) {
+      date = new ExprDateValue(date.stringValue());
+    }
+    return new ExprIntegerValue(date.dateValue().getDayOfWeek().getValue() == 7 ? 1
+        : date.dateValue().getDayOfWeek().getValue() + 1);
+  }
+
+  /**
+   * Day of Year implementation for ExprValue.
+   * @param date ExprValue of Date type.
+   * @return ExprValue.
+   */
+  private ExprValue exprDayOfYear(ExprValue date) {
+    if (date instanceof ExprStringValue) {
+      date = new ExprDateValue(date.stringValue());
+    }
+    return new ExprIntegerValue(date.dateValue().getDayOfYear());
+  }
+
+  /** From_days implementation for ExprValue.
+   * @param exprValue Day number N.
+   * @return ExprValue.
+   */
+  private ExprValue exprFromDays(ExprValue exprValue) {
+    return new ExprDateValue(LocalDate.ofEpochDay(exprValue.longValue() - DAYS_0000_TO_1970));
+  }
+
+  /**
+   * Hour implementation for ExprValue.
+   * @param exprValue ExprValue of Time type.
+   * @return ExprValue.
+   */
+  private ExprValue exprHour(ExprValue exprValue) {
+    return new ExprIntegerValue(exprValue.timeValue().getHour());
+  }
+
+  /**
+   * Microsecond implementation for ExprValue.
+   * @param exprValue ExprValue of Time type.
+   * @return ExprValue.
+   */
+  private ExprValue exprMicrosecond(ExprValue exprValue) {
+    return new ExprIntegerValue(
+        TimeUnit.MICROSECONDS.convert(exprValue.timeValue().getNano(), TimeUnit.NANOSECONDS));
+  }
+
+  /**
+   * Minute implementation for ExprValue.
+   * @param exprValue ExprValue of Time type.
+   * @return ExprValue.
+   */
+  private ExprValue exprMinute(ExprValue exprValue) {
+    return new ExprIntegerValue(exprValue.timeValue().getMinute());
+  }
+
+  /**
+   * Month for date implementation for ExprValue.
+   * @param date ExprValue of Date type.
+   * @return ExprValue.
+   */
+  private ExprValue exprMonth(ExprValue date) {
+    if (date instanceof ExprStringValue) {
+      date = new ExprDateValue(date.stringValue());
+    }
+    return new ExprIntegerValue(date.dateValue().getMonthValue());
+  }
+
+  /**
+   * Name of the Month implementation for ExprValue.
+   * @param date ExprValue of Date type.
+   * @return ExprValue.
+   */
+  private ExprValue exprMonthName(ExprValue date) {
+    if (date instanceof ExprStringValue) {
+      date = new ExprDateValue(date.stringValue());
+    }
+    return new ExprStringValue(
+        date.dateValue().getMonth().getDisplayName(TextStyle.FULL, Locale.getDefault()));
+  }
+
+  /**
+   * Quarter for date implementation for ExprValue.
+   *
+   * @param date ExprValue of Date type.
+   * @return ExprValue.
+   */
+  private ExprValue exprQuarter(ExprValue date) {
+    if (date instanceof ExprStringValue) {
+      date = new ExprDateValue(date.stringValue());
+    }
+    return new ExprIntegerValue((date.dateValue().getMonthValue() % 3) == 0
```

Comment:
Should latch variable and reuse since it's a lot of function calls.

int month = date.dateValue().getMonthValue();
return new ExprIntegerValue((month / 3) + ((month % 3) == 0 ? 0 : 1));

Reply:
Thanks, I added a variable.